### ISSUE TITLE
defaultTLSAlloc() returns NULL in Breath of the Wild

### ISSUE THREAD

**Issue by CarlKenner** [decaf-emu/decaf-emu#401]
sTLSAlloc() is called before OSDynLoad_SetTLSAllocator().

internal::dynLoadTLSAlloc() returns NULL to tls_get_addr(), causing an access violation trying to write to NULL in the next line:
std::memset(&thread->tlsSections[oldCount], 0, sizeof(OSTLSSection) * (thread->tlsSectionCount - oldCount));

sTLSAlloc(8, 4) calls defaultTLSAlloc(8, 4), which calls allocFromDefaultHeapEx(8, 4) which calls defaultAllocFromDefaultHeapEx(8, 4), which calls MEMAllocFromExpHeapEx(MEM2, 8, 4).

freeList.head is NULL, so MEMAllocFromExpHeapEx never loops or finds a block, so it never sets newBlock, which is still NULL.
It checks that newBlock is NULL, and calls MEMDumpHeap(MEM2->header), then returns NULL.

MEMDumpHeap calls dumpExpandedHeap().

Note: The heap is created successfully (with freeList.head set), and some other memory is allocated successfully before this.

**Comment by exjam** [comment]
It's because the game allocates the whole default heap, then tries to do tls_get_addr which fails as it  tries to allocate from the default heap.

So it basically does
```
auto heap = MEMGetBaseHeapHandle(MEM2)
auto size = MEMGetAllocatableSizeForExpHeapEx(heap)
MEMAllocFromExpHeapEx(heap, size);
...
tls_get_addr(x);
```

It's possible if the game is importing OSDynLoad_SetTLSAllocator then maybe we are missing some code path which causes the game calls this. Whether it's some sort of initialisation function being missed or whether we are returning an incorrect value from a function leading to this.

**Comment by CarlKenner** [comment]
I just discovered that.

We only need 8 bytes + sizeof(MEMExpHeapBlock). Should we change MEMGetAllocatableSizeForExpHeapEx to reserve that much memory?

Or is it supposed to allocate TLS from somewhere other than MEM2? (MEM1?)

BTW, how do I pass options other than "play" to the exe? And how do I find the log file when I edited the config file to enable logging to a file?

**Comment by VelocityRa** [comment]
@CarlKenner Every other arg goes _after_ `play <game_path>`.
Log files are in `decaf-emu\build` for me.

**Comment by exjam** [comment]
It is not correct to arbitrarily reserve memory in MEMGetAllocatableSizeForExpHeapEx. That would be called a game specific hack which is a no no 💩 . Also it would be more than 8+sizeof(block) because you'd need to support all further TLS allocations too, not just this first one which is preventing the game from running further.

The dynload TLS allocator defaults to the default heap. You'll need to reverse coreinit.rpl if you want to verify decaf's implementation of this.

Log files are output in the working directory.

Arguments can come anywhere in the command line, play is not really an "argument". All arguments are preceded by --.

eg the following would work:
decaf.exe --arg1 play --arg2 <path> --arg3
decaf.exe --arg1 --arg2 play <path>
decaf.exe play <path> --arg1 --arg2 --arg3

**Comment by CarlKenner** [comment]
The game is allocating TLS after allocating all the allocatable size. So either it expects there to be holes in the memory, or MEMGetAllocatableSizeForExpHeapEx() doesn't return the whole memory, or TLS memory is supposed to be allocated from somewhere else, or the game is supposed to free some of the memory it allocated, or the TLS memory is supposed to be allocated using some other heap system within the huge block that it allocated.

I haven't tested enough games to know if this issue is unique to Breath of the Wild.

There's no log file in my working directory.

**Comment by CarlKenner** [comment]
I think it must be messing with the sbrk and __ghs_alloc functions, or something like that. It must be trying to allocate TLS from the memory it allocated, but I don't know how.

**Comment by exjam** [comment]
Neither sbrk or __ghs_alloc is imported from coreinit in botw!

I highly recommend you actually reverse the game + coreinit using IDA or something rather than just trying to guess from the log files / decaf source. You will get a much clearer picture of what's going on!

For example, here's proof that TLS allocations come from the default heap:

__tls_get_addr calls the allocator set by OSDynLoad_SetTLSAllocator:
https://puu.sh/uZXVB/8a30fecb6a.png

CoreInitDefaultHeap sets a default allocator for OSDynLoad_SetTLSAllocator:
https://puu.sh/uZXXM/4590ea0cd5.png

The default allocator calls MEMAllocFromDefaultHeapEx:
https://puu.sh/uZXYG/e4d14aa967.png

MEMAllocFromDefaultHeapEx is also initialised in CoreInitDefaultHeap:
https://puu.sh/uZYak/639d080551.png

And it's a very simple function which just calls MEMAllocFromHeapEx:
https://puu.sh/uZYbm/600c82b0a6.png

Which uses a heap initialised with MEMCreateExpHeapEx from a sub-function of CoreInitDefaultHeap:
https://puu.sh/uZYdx/03f000ce46.png

Which you will see is called with the memory bounds returned from OSGetMemBound:
https://puu.sh/uZYdQ/b839780e88.png

You can also go look at the implementation of MEMGetAllocatableSizeForExpHeapEx and see it does not reserve any memory.

So yeah 5 mins in IDA gives a lot of useful info!

**Comment by CarlKenner** [comment]
I guess the problem is that we're supposed to allocate all the TLS sections before calling __preinit_user. 
It's failing on the first TLS module.

**Comment by exjam** [comment]
You can't allocate it early,, it's a thread local storage - it's allocated per thread, so you don't know what to allocate unless you magically know the number of threads beforehand. In which case you probably want to look in OSCreateThread for any hint of initialising the thread's TLS storage from some pre-allocated area.

I feel that's very unlikely though, otherwise why bother putting the allocation code in __tls_get_addr which uses the dynload tls allocator?

**Comment by CarlKenner** [comment]
This bug has me stumped for now. I think I'll just give up on it for a while.

**Comment by achurch** [comment]
FYI, I was messing with BotW a bit the other day and it looks like it tries to allocate just a bit more memory than Decaf makes available (IIRC it tries to allocate a fixed size of 0x36400000 when only 0x363nnnnn is available), which makes me wonder if real hardware arranges memory differently so that >=0x36400000 (or whatever amount) is in fact available at that point. If I expand MEM2 by a bit, the game does successfully boot as far as the start screen.

**Comment by brett19** [comment]
Hey Guys,
It is worth pointing out that when I built the memory mapping section, it is not 100% accurate with how the system behaves, specifically there is a lot of logic about how system/user libraries get loaded and how memory is initial set up when the application starts which getting perfect parity with the production system was a challenge.  Also note that we arbitrarily allocate a chunk of MEM2 for system startup which was done to reflect how DK2 expects a certain upper limit of Memory available before it assumes your on a development system (which has 2GB rather than 1GB).
Cheers, Brett

**Comment by ds84182** [comment]
Perhaps botw allocates more memory than a usual game. Whenever you exit the
home menu and go back you have to wait for it to load before you can
continue. This affects the title screen also.

**Comment by achurch** [comment]
Just to mention, I do have a patch on my local branch that makes the size of MEM2 configurable: https://github.com/achurch/decaf-emu/commit/60ed3dad0a8a7c06cf11985db9547326fe77a44a
Dunno if it would be appropriate for master, but could be convenient for working around this for now (or if you want to try emulating a devkit).

**Comment by exjam** [comment]
It's probably related to the call to OSEnableOverlayArena.

The interesting thing is that OSEnableOverlayArena is meant to return addr & size as arguments but zelda just completely ignores that, so it must be allocated to a predictable address - or merged with the default heap / foreground heap.

It does seem to allocate the FG heap right after enabling the overlay arena:
```
[14:38.506676 debug:w108] 0x02F271E8 OSIsEnabledOverlayArena()
[14:38.506742 debug:w108] 0x02F27240 OSEnableOverlayArena(0x1, 0x010CCFA8, 0x010CCFAC)
[14:38.508034 debug:w108] 0x02F1867C OSGetSystemTime()
[14:38.508323 debug:w108] 0x03064738 OSGetSystemTime()
[14:38.508508 debug:w108] 0x02F24ECC MEMGetBaseHeapHandle(8)
[14:38.508569 debug:w108] 0x02F24ED8 MEMGetAllocatableSizeForFrmHeapEx(0xE0000000, 4)
[14:38.508629 debug:w108] 0x02F24EF4 MEMAllocFromFrmHeapEx(0xE0000000, 0x027FFFB4, 4)
```

This led to me prodding at the Wii U using OSEffectiveToPhysical to map out our memory regions:

```
Physical:
0x00000000 - 0x01FFFFFF MEM1
0x10000000 - 0x8FFFFFFF MEM2

Virtual Address Space:
0x01000000 - 0x017fffff maps to 0x32000000 - 0x327fffff
0x0D000000 - 0x0fffffff maps to 0x8d000000 - 0x8fffffff
0x10000000 - 0x4CF9ffff maps to 0x50000000 - 0x8cf9ffff
0xe0000000 - 0xe3ffffff maps to 0x14000000 - 0x17ffffff
0xf4000000 - 0xf5ffffff maps to 0x00000000 - 0x01FFFFFF
0xf6000000 - 0xf67fffff maps to 0x1b000000 - 0x1b7fffff
0xf8000000 - 0xfaffffff maps to 0x18000000 - 0x1affffff
0xfb000000 - 0xfb7fffff maps to 0x1c800000 - 0x1cffffff
0xfc000000 - 0xfc3fffff maps to 0x0c300000 - 0x0c3fffff
0xfd000000 - ?          maps to 0xfd000000 - ?
0xfe000000 - 0xfe7fffff maps to 0x1c000000 - 0x1c7fffff
```

Which is similar to http://wiiubrew.org/wiki/Physical_Memory and http://wiiubrew.org/wiki/Cafe_OS#Virtual_Memory_Map

The process my .rpx was running as didn't seem to have permission to enable the overlay arena so I couldn't figure out where that was.

I think we need to change our memory setup to more closely match the consoles.

**Comment by exjam** [comment]
Well I merged my new virtual / physical memory implementation to master - it doesn't fix the Zelda issue because we are still not giving the game enough memory. I'd bet it's something to do with the overlay arena.

**Comment by achurch** [comment]
FWIW, here's the rough call sequence (addresses are from J version):
```
0x2b43c40 - store 0x3b600000 (desired alloc size) to *(r1+8)
0x2b43c50 - call OSGetMemBound(2) (returns *addr=0x10000000, *size=0x3c900000)
0x2b43c58 - call 0x3063874(r1+8)
  0x3063880 - call 0x2f185b8(r3) (trampoline to 0x2f1bd60)
    0x2f1bd70 - call 0x2f1b90c(r3)
      0x2f1b950 - call 0x2f269b4(*r3)  // *r3 == 0x3b600000
        0x2f269ec - call 0x2f5a170($DATA+0x4cf704, 0x3b600000)
          0x2f5a190 - call MEMGetBaseHeapHandle(1)  // returns 0x10000000
          0x2f5a198 - call MEMGetAllocatableSizeForExpHeapEx(0x10000000, 4)  // returns 0x3b39f1cc
          0x2f5a19c - set r3 = min(allocatableSize, r3)  // clamps desired size to allocatableSize
          0x2f5a1c0 - call MEMAllocFromDefaultHeapEx(r3)
```
And when the mutex alloc fails, Decaf logs this:
```
[29:29.487308 debug:w108] MEMExpHeap(0x10000000)
[29:29.487315 debug:w108] Status Address   Size       Group
[29:29.487317 debug:w108] USED  0x11560e20 0x3b39f1cc 0
[29:29.487329 debug:w108] USED  0x10560e0c 0x 1000000 0
[29:29.487330 debug:w108] USED  0x100000ec 0x  560d0c 0
```
I've never looked at the overlay arena so I can't say anything about it, but it looks like there's just not enough free space in MEM2.

I do note that 0x40000000 - 0x3c900000 = 0x3700000 which is fairly close to the uncompressed ELF size (0x323dd00), but the binutils "size" tool says the actual code+data+bss size is 0x26f7b3d, and 0x323dd00 - 0x26f7b3d ~= 0xb00000 which would be enough to allow the 0x3b600000-byte allocation to succeed with enough space left over for mutexes and what-not. I haven't looked at the loader in detail, but is it possible the kernel is keeping things like relocation sections in memory when it should be discarding them after the load is complete?

**Comment by exjam** [comment]
**tldr: The fix is at the bottom of this comment**

Well we allocate an area for code as specified in the cos.xml:
 `<max_codesize type="hexBinary" length="4">03700000</max_codesize>`

Zelda wants 0x36400000  memory so that leaves us with 0x40000000 - 0x36400000 = 0x09C00000 bytes to play with.

Then we take off the max_codesize from the cos.xml 0x09C00000 - 0x03700000 = 0x06500000.

So we have 0x06500000 for data sections, looking at the section dump from below we can see the last section, .thrbss is 0xc big at 0x10560E00, so that means our data is (0x10560E00+0xC)-0x10000000 bytes big which is 0x00560E0C.

0x06500000 - 0x00560E0C = 0x05F9F1F4, so we should have plenty of memory left over...!

```
[47:14.063831 debug:w1FF] Section Summary:
  [Nr] Name                 Type             Addr     Off      Size     ES Flag  Lk Info Align VAddr    VSize   
  [ 0]                      SHT_NULL         00000000 00000000 00000000 00        0    0     0 00000000 00000000
  [ 1] .syscall             SHT_PROGBITS     02000000 0013D600 00000008 00    AX  0    0    32 02000000 00000008
  [ 2] .text                SHT_PROGBITS     02000020 0013D640 00F03533 00   AXZ  0    0    32 02000020 02195FC4
  [ 3] .rodata              SHT_PROGBITS     10000000 00000780 00134891 00   WAZ  0    0    32 10000100 003F9040
  [ 4] .data                SHT_PROGBITS     103F9040 00135040 000068C7 00   WAZ  0    0    32 103F9140 000211CC
  [ 5] .module_id           SHT_PROGBITS     1041A220 0013B940 0000008B 00   WAZ  0    0    32 1041A320 000000E0
  [ 6] .bss                 SHT_NOBITS       1041A300 00000000 001469E1 00    WA  0    0   256 1041A400 001469E1
  [ 7] .thrbss              SHT_NOBITS       10560D00 00000000 0000000C 00   WAT  0    0    32 10560E00 0000000C
  [ 8] .rela.text           SHT_RELA         00000000 01040B80 0038A7E7 0C     Z 30    2     4 00000000 00000000
  [ 9] .rela.rodata         SHT_RELA         00000000 013CB380 00133664 0C     Z 30    3     4 00000000 00000000
  [10] .rela.data           SHT_RELA         00000000 014FEA00 000077AF 0C     Z 30    4     4 00000000 00000000
  [11] .fimport_gx2         SHT_RPL_IMPORTS  C0004480 0013BA00 00000020 00   AXZ  0    0     4 E6000000 00000418
  [12] .fimport_snd_core    SHT_RPL_IMPORTS  C00048C0 0013BA40 00000021 00   AXZ  0    0     4 E6000418 000001F0
  [13] .fimport_snd_user    SHT_RPL_IMPORTS  C0004AC0 0013BA80 00000018 00    AX  0    0     4 E6000608 00000018
  [14] .fimport_h264        SHT_RPL_IMPORTS  C0004B00 0013BAC0 0000001B 00   AXZ  0    0     4 E6000620 00000070
  [15] .fimport_nn_save     SHT_RPL_IMPORTS  C0004B80 0013BB00 0000001E 00   AXZ  0    0     4 E6000690 00000048
  [16] .fimport_vpad        SHT_RPL_IMPORTS  C0004C00 0013BB40 0000001B 00   AXZ  0    0     4 E60006D8 00000058
  [17] .fimport_vpadbase    SHT_RPL_IMPORTS  C0004C80 0013BB80 00000018 00    AX  0    0     4 E6000730 00000018
  [18] .fimport_proc_ui     SHT_RPL_IMPORTS  C0004CC0 0013BBC0 0000001E 00   AXZ  0    0     4 E6000748 00000030
  [19] .fimport_padscore    SHT_RPL_IMPORTS  C0004D00 0013BC00 0000001F 00   AXZ  0    0     4 E6000778 00000068
  [20] .fimport_coreinit    SHT_RPL_IMPORTS  C0004D80 0013BC40 00000026 00   AXZ  0    0     4 E60007E0 000004C0
  [21] .dimport_coreinit    SHT_RPL_IMPORTS  C0005240 0013BC80 0000001F 00    AZ  0    0     4 E6000CA0 00000048
  [22] .fimport_sysapp      SHT_RPL_IMPORTS  C00052C0 0013BCC0 0000001C 00   AXZ  0    0     4 E6000CE8 00000020
  [23] .fimport_nn_act      SHT_RPL_IMPORTS  C0005300 0013BD00 0000001D 00   AXZ  0    0     4 E6000D08 00000020
  [24] .dimport_nn_act      SHT_RPL_IMPORTS  C0005340 0013BD40 00000010 00     A  0    0     4 E6000D28 00000010
  [25] .fimport_nn_boss     SHT_RPL_IMPORTS  C0005380 0013BD80 0000001E 00   AXZ  0    0     4 E6000D38 00000090
  [26] .dimport_nn_boss     SHT_RPL_IMPORTS  C0005440 0013BDC0 00000010 00     A  0    0     4 E6000DC8 00000010
  [27] .fimport_nn_nfp      SHT_RPL_IMPORTS  C0005480 0013BE00 0000001D 00   AXZ  0    0     4 E6000DD8 000000A0
  [28] .fimport_nsysnet     SHT_RPL_IMPORTS  C0005540 0013BE40 00000010 00    AX  0    0     4 E6000E78 00000010
  [29] .fimport_nn_ac       SHT_RPL_IMPORTS  C0005580 0013BE80 00000010 00    AX  0    0     4 E6000E88 00000010
  [30] .symtab              SHT_SYMTAB       C0000000 0013BEC0 000009C4 10    AZ 31 1076042     4 E6000E98 00001DD0
  [31] .strtab              SHT_STRTAB       C0001DD0 0013C8C0 00000C65 00    AZ  0    0     1 E6002C68 000024DB
  [32] .shstrtab            SHT_STRTAB       C00042AB 0013D540 000000B0 00    AZ  0    0     1 E6005143 000001AC
  [33]                      SHT_RPL_CRCS     00000000 000005B8 0000008C 04        0    0     4 00000000 00000000
  [34]                      SHT_RPL_FILEINFO 00000000 00000680 000000DD 00        0    0     4 00000000 00000000
```

So as you can see from your MEMExpHeap dump in the log there is an odd looking allocation in there
```
[29:29.487317 debug:w108] USED  0x11560e20 0x3b39f1cc 0 <--- this is what zelda allocated
[29:29.487329 debug:w108] USED  0x10560e0c 0x 1000000 0 <--- what is this?????
[29:29.487330 debug:w108] USED  0x100000ec 0x  560d0c 0 <--- this matches with the .data sizes above
```

If you grep the code for 0x1000000 you can find an arbitrary allocation we have leftover from a long time ago like @brett19  mentioned,

https://github.com/decaf-emu/decaf-emu/blob/f08bcca97be70541f57d383215252d698b2fe2ed/src/libdecaf/src/modules/coreinit/coreinit_scheduler.cpp#L670-L672

Remove that and Zelda gets enough memory.

Although now we have a new problem where it looks like Zelda expects the overlay arena to be allocated at virtual address 0xA0000000 where I currently have it at 0x60000000. It's funny though because 0xA0000000 - 0xDFFFFFFF is the memory range available to the OSAllocVirtAddr family of functions. It's possible that when the overlay arena is enabled then it utilises a part of this virtual address to map in the overlay arena. I will do some more investigations and fix it!

**Comment by exjam** [comment]
Oh right - looks like I misread "lis -0x6000" as 0x60000000 but it's actually 0xA0000000, damn you IDA! So yeah it maps the overlay arena into that address space.
